### PR TITLE
support HyperModel.fit()

### PR DESCRIPTION
**Pull request by haifeng-jin**
Implement `HyperModel.fit()` according to the [design doc](https://github.com/haifeng-jin/governance/blob/kt/rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md#the-hypermodel-workflow).

### CODE REVIEW

File: keras_tuner/engine/hypermodel.py [hunk 1]
```diff
@@ -80,6 +80,22 @@ def _build_wrapper(self, hp, *args, **kwargs):
...
+    def fit(self, hp, model, *args, **kwargs):
+        """Train the model.
+
+        Args:
+            hp: HyperParameters.
+            model: `keras.Model` built in the `build()` function.
+            **kwargs: Anything the user defines. They are passed from
+                `Tuner.search()`. if contains `callbacks`, a list of prebuild
+                Keras callbacks for model checkpointing and tensorboard
+                configuration will be added to callbacks provided by the user.
```
Comment (fchollet) [inline keras_tuner/engine/hypermodel.py:92]: `callbacks` is always passed by `search()`, right? If the user passed a callbacks list, it gets appended to, and if no callbacks list is passed, one gets created?

File: keras_tuner/engine/hypermodel.py [hunk 2]
```diff
@@ -80,6 +80,22 @@ def _build_wrapper(self, hp, *args, **kwargs):
...
+
+        Returns:
+            A `History` object, a similar dictionary, or a single value.
```
Comment (fchollet) [inline keras_tuner/engine/hypermodel.py:95]: Please provide more details about the dict and single value cases. Also mention that `History` instance is the output of `model.fit()`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_tuner/engine/hypermodel.py (hunk 1 above)
Comment (haifeng-jin) [reply keras_tuner/engine/hypermodel.py]: Yes, the description is revised.

File: keras_tuner/engine/hypermodel.py [hunk 3]
```diff
@@ -47,6 +51,42 @@ def build(self, hp):
...
+
+    ```python
+    class MyHyperModel(kt.HyperModel):
+        def build(self, hp):
+            ...
+
+        def fit(self, hp, model, *args, **kwargs):
+            return model.fit(
+                *args,
+                epochs=hp.Int("epochs", 5, 20)
```
Comment (fchollet) [inline keras_tuner/engine/hypermodel.py:66]: Missing comma

**Review by fchollet — APPROVED** [review]
LGTM, thanks for the update

**Pull request merged**